### PR TITLE
Kiali Helm chart - adds secret from random digits if no password is set 

### PR DESCRIPTION
**Pull request by monotek**
* Creates a kiali password from a random 40 digit number if no password is set
* also creates the kubernetes secret containing the password (removes install step 5 from the istio readme)

### CODE REVIEW

File: install/kubernetes/helm/istio/Chart.yaml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 apiVersion: v1
 name: istio
-version: 1.1.0
+version: 1.1.1
```
Comment (sdake) [inline install/kubernetes/helm/istio/Chart.yaml]: the next istio release is 1.1.0 - please revert this change

File: install/kubernetes/helm/istio/values.yaml [hunk 2]
```diff
@@ -78,6 +78,10 @@ tracing:
 #
 kiali:
   enabled: false
+  # username & password for kiali
+  # if no password is set a 40 digit random password will be set
+  username: admin
+  passphrase: ""
```
Comment (sdake) [inline install/kubernetes/helm/istio/values.yaml]: please place this in the charts/kiali/values.yaml.  Only globals, and enabled/disabled defaults should be listed in values.yaml.  This isn't documented anywhere in our repo - so no harm 👍

File: install/kubernetes/helm/istio/Chart.yaml (hunk 1 above)
Comment (monotek) [reply install/kubernetes/helm/istio/Chart.yaml]: OK. But how do you track changes on the chart?

**Comment by monotek** [comment]
I signed it!

File: install/kubernetes/helm/istio/charts/kiali/values.yaml [hunk 3]
```diff
@@ -19,9 +19,11 @@ ingress:
     #   hosts:
     #     - kiali.local
 dashboard:
+  # if not secretname is set kiali fullname will be used
   secretName: kiali
+  # if no password is set a 40 digit random password will be set
   usernameKey: username
-  passphraseKey: passphrase
```
[the remark this answers is not in the document]
Comment (monotek) [reply install/kubernetes/helm/istio/charts/kiali/values.yaml]: Thanks for the hint. Added new values "username" & "passphrase".

File: install/kubernetes/helm/istio/charts/kiali/values.yaml [hunk 4]
```diff
@@ -19,9 +19,11 @@ ingress:
     #   hosts:
     #     - kiali.local
 dashboard:
+  # if not secretname is set kiali fullname will be used
   secretName: kiali
+  # if no password is set a 40 digit random password will be set
```
[the remark this answers is not in the document]
Comment (monotek) [reply install/kubernetes/helm/istio/charts/kiali/values.yaml]: Thanks for the hint. Added new values "username" & "passphrase".

File: install/kubernetes/helm/istio/charts/kiali/values.yaml [hunk 5]
```diff
@@ -19,9 +19,13 @@ ingress:
     #   hosts:
     #     - kiali.local
 dashboard:
+  # if not secretname is set kiali fullname will be used
   secretName: kiali
   usernameKey: username
   passphraseKey: passphrase
+  # if no password is set a 40 digit random password will be set
+  username: admin
```
Comment (gyliu513) [inline install/kubernetes/helm/istio/charts/kiali/values.yaml]: We do not want to expose username and password in values.yaml because of security issue, please refer to https://github.com/istio/istio/issues/7447 for detail

File: install/kubernetes/helm/istio/values.yaml (hunk 2 above)
Comment (sdake) [reply install/kubernetes/helm/istio/values.yaml]: @monotek ya - i'd like to get rid of our custom implementations of Grafana and Prometheus, but I'm not sure if it is possible.  We have direct integration in our own charts to work well with Istio.  Before we can do any of that, we need `helm dep udpate` working - see PR: https://github.com/istio/istio/pull/9306

File: install/kubernetes/helm/istio/Chart.yaml (hunk 1 above)
Comment (sdake) [reply install/kubernetes/helm/istio/Chart.yaml]: Per release of Istio, Test and Relaese (T&R) seds this file to the next major version (1.2.0).  We don't track individual changes -1,-2, etc as Helm doesn't have any mechanism of templating version fields.  As we have a slew of version fields, tracking individual changes is not tenable.

File: install/kubernetes/helm/istio/charts/kiali/values.yaml (hunk 5 above)
Comment (sdake) [reply install/kubernetes/helm/istio/charts/kiali/values.yaml]: I think the short answer is you don't do this for either username or password.  We have moved to reference by key names for the passwords and usernames throughout the codebase for security reasons.  If you need passwords specified, doing so in the Helm `values.yaml` file is the wrong place to do it, as it exposes the secrets during helm workflows (via debug and key resolution).

Instead i'd recommend sorting out a workflow which creates your secrets (username and passphrase) via kubectl secret create.

Comment (monotek) [reply install/kubernetes/helm/istio/charts/kiali/values.yaml]: We don't store our secrets in the values.yaml directly. We use an additional secrets.yaml (encrypted by SOPS). In additon we're using the Helm secrets plugin. 

To be able to use the chart, without the need to use a SOPS encrypted secrets.yaml (like for CI), we use empty password values which are replaced with random strings by default.

Only our jenkins user is able to decrypt secrets. Devs only have the possbility to encrypt (via Google KMS). So nobody would see a real password while debugging locally. You just get the random string password in helm template / debug. "helm get" would work of cause but i think you could also get it via "kubectl get secret" if you're able to create the secret via kubectl so i'm not sure if there is a real difference security wise?

As our setup is maybe not so common i understand your point and i guess this pr has no chance to be merged, right?

Comment (sdake) [reply install/kubernetes/helm/istio/charts/kiali/values.yaml]: Kubernetes secrets can be secured (and encrypted at rest as well).  See https://kubernetes.io/docs/concepts/configuration/secret/#best-practices - list/view for information on locking out kubectl get/list, so there is a delta security wise.

Regarding your setup, Id receomment an "istio-secrets" helm chart which registers your secrets to the defaults you specify for usernameKey and passphraseKey.  This is not secure, but it would work with your workflow.  Once a secret leaks once, it is forever tainted.

With this model the current workflow used with SOPS+helm secret plugin to create a simple chart that only registers the passwords (IOW outside of the Istio  
charts).

We really want Istio to be super secure, and Kubernetes secrets offer the most secure path (hence why the default passwords were removed from the values.yaml - which cannot be encrypted at rest and can be read by multiple system tools, including the console during rendering.

**Comment by monotek** [comment]
closing...
